controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1420,6 +1486,89 @@ private TxnStatus handleDataNotFoundException(Throwable ex) {
+                .thenCompose(record -> {
+                    if (record == null) {
+                        // Attempt to create a new record. It is possible that while we are attempting to create the record, 
+                        // a concurrent request could have created it. 
+                        // For example: a writer sends a request to one of controller instances to note writer mark and 
+                        // the connection is dropped. It could then send the request to noteTime to a different controller 
+                        // instance. So two controller instances could attempt concurrent creation of mark for the said writer.
+                        // In this case, we may have found mark to be inexistent when we did a get but it may have been created 
+                        // when we attempt to create it. So immediately after attempting a create, if we get DataExists, 
+                        // we will call noteWriterMark recursively. 
+                        return Futures.exceptionallyComposeExpecting(
+                                createWriterMarkRecord(writer, timestamp, newPosition)
+                                        .thenApply(v -> WriterTimestampResponse.SUCCESS),
+                                e -> Exceptions.unwrap(e) instanceof StoreException.DataExistsException, 
+                                () -> noteWriterMark(writer, timestamp, position));
+                    } else {
+                        // sanity check and update
+                        if (record.getObject().getTimestamp() > timestamp) {
+                            // existing time is already ahead of new time
+                            return CompletableFuture.completedFuture(WriterTimestampResponse.INVALID_TIME);
+                        }
+                        if (!compareWriterPositions(record.getObject().getPosition(), newPosition)) {
+                            // existing position is already ahead of new position
+                            return CompletableFuture.completedFuture(WriterTimestampResponse.INVALID_POSITION);
+                        }
+                        
+                        // its a valid mark, update it
+                        return updateWriterMarkRecord(writer, timestamp, newPosition, record.getVersion())
+                                .thenApply(v -> WriterTimestampResponse.SUCCESS);
+                    }
+                });
+    }
+
+
+    /**
+     * Compares two given positions and returns true if position 2 is ahead of position 1. 
+     * 
+     * Note: This is not absolutely correct implementation because that would be a very costly computation. 
+     * This method simply checks if position 2 has a higher max segment number than position 1. 
+     * If those are equal, then it simply checks the offsets in each segment in position 2 to be ahead of same segment in 
```

Comment:
The implementation of the method actually is stronger than this comment. It compares all segment that overlap have increased.

Reply:
The comment mentions that
>       * If those are equal, then it simply checks the offsets in each segment in position 2 to be ahead of same segment in

Reply:
Right, it doesn't just do that check "if they are equal" it does it all the time.

Reply:
Fixed the impl to do the latter iff they are equal

Reply:
That's the opposite of what I was suggesting. The impl as it was was fine. I just wanted the doc to reflect it.

Reply:
I replied to the other comment, the implementation was doing unnecessary sorting and not making use of it. I changed that. And as for two conditions that are being checked, if first condition fails, we no longer check second condition. the java doc reflects that.